src/compiler/checker.ts

```diff
@@ -25203,6 +25203,48 @@ namespace ts {
             return literalType;
         }
 
+        /**
+         * Replaces all references to structurally equivalent object and array literal types in the given list
+         * with references to a single one of those types. The process is applied recursively to properties of
+         * object literals and elements of array literals.
+         */
+        function deduplicateObjectOrArrayLiteralTypes(types: Type[]) {
+            if (!some(types, isObjectOrArrayLiteralType)) {
+                return types;
+            }
+            const typeMap = new Map<string, Type>();
+            return sameMap(types, getInternedType);
+
+            function getInternedType(type: Type): Type {
+                if (type.flags & TypeFlags.Union) {
+                    const newTypes = sameMap((<UnionType>type).types, getInternedType);
+                    return newTypes !== (<UnionType>type).types ? getUnionType(newTypes) : type;
+                }
+                const key = getLiteralTypeKey(type);
+                return key ? typeMap.get(key) || (typeMap.set(key, type), type) : type;
+            }
+
+            function getLiteralTypeKey(type: Type): string | undefined {
+                const objectFlags = getObjectFlags(type);
+                if (objectFlags & ObjectFlags.ObjectLiteral) {
+                    const props = getPropertiesOfObjectType(type);
+                    const propTypes = map(props, p => getInternedType(getTypeOfSymbol(p)));
+                    const nameLengths = map(props, p => p.flags & SymbolFlags.Optional ? -(<string>p.escapedName).length : (<string>p.escapedName).length);

```

Comment:
As-is, I think this'd intern `{ set a(x: string) {} }` and `{ get a(){ return ""; } }` to the same type, even though the later has a `readonly a`. Maybe just encoding the symbol flags wholesale in the key will be the most future proof thing, and then you can just join the names with `,` and not worry about the name length list thing.

Reply:
Yeah, I'll look to make sure we properly preserve that distinction. The name length list is still needed since property names can contain any character, including any separator we might use to separate them.